File: core/field_bool_test.go
package core_test

import (
	"context"
	"fmt"
	"testing"

	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tests"
)

func TestBoolFieldBaseMethods(t *testing.T) {
	testFieldBaseMethods(t, core.FieldTypeBool)
}

func TestBoolFieldColumnType(t *testing.T) {
	t.Parallel()

	tests.DualDBTest(t, func(t *testing.T, app *tests.TestApp, dbType tests.DBType) {
		f := &core.BoolField{}

		expected := "BOOLEAN DEFAULT FALSE NOT NULL"

		if v := f.ColumnType(app); v != expected {
			t.Fatalf("Expected\n%q\ngot\n%q", expected, v)
		}
	})
}

func TestBoolFieldPrepareValue(t *testing.T) {
	t.Parallel()

	tests.DualDBTest(t, func(t *testing.T, app *tests.TestApp, dbType tests.DBType) {
		f := &core.BoolField{}
		record := core.NewRecord(core.NewBaseCollection("test"))

		scenarios := []struct {
			raw      any
			expected bool
		}{
			{"", false},
			{"f", false},
			{"t", true},
			{1, true},
			{0, false},
		}

		for i, s := range scenarios {
			t.Run(fmt.Sprintf("%d_%#v", i, s.raw), func(t *testing.T) {
				v, err := f.PrepareValue(record, s.raw)
				if err != nil {
					t.Fatal(err)
				}

				if v != s.expected {
					t.Fatalf("Expected %v, got %v", s.expected, v)
				}
			})
		}
	})
}

func TestBoolFieldValidateValue(t *testing.T) {
	t.Parallel()

	tests.DualDBTest(t, func(t *testing.T, app *tests.TestApp, dbType tests.DBType) {
		collection := core.NewBaseCollection("test_collection")

		scenarios := []struct {
			name        string
			field       *core.BoolField
			record      func() *core.Record
			expectError bool
		}{
			{
				"invalid raw value",
				&core.BoolField{Name: "test"},
				func() *core.Record {
					record := core.NewRecord(collection)
					record.SetRaw("test", 123)
					return record
				},
				true,
			},
			{
				"missing field value (non-required)",
				&core.BoolField{Name: "test"},
				func() *core.Record {
					record := core.NewRecord(collection)
					record.SetRaw("abc", true)
					return record
				},
				true, // because of failed nil.(bool) cast
			},
			{
				"missing field value (required)",
				&core.BoolField{Name: "test", Required: true},
				func() *core.Record {
					record := core.NewRecord(collection)
					record.SetRaw("abc", true)
					return record
				},
				true,
			},
			{
				"false field value (non-required)",
				&core.BoolField{Name: "test"},
				func() *core.Record {
					record := core.NewRecord(collection)
					record.SetRaw("test", false)
					return record
				},
				false,
			},
			{
				"false field value (required)",
				&core.BoolField{Name: "test", Required: true},
				func() *core.Record {
					record := core.NewRecord(collection)
					record.SetRaw("test", false)
					return record
				},
				true,
			},
			{
				"true field value (required)",
				&core.BoolField{Name: "test", Required: true},
				func() *core.Record {
					record := core.NewRecord(collection)
					record.SetRaw("test", true)
					return record
				},
				false,
			},
		}

		for _, s := range scenarios {
			t.Run(s.name, func(t *testing.T) {
				err := s.field.ValidateValue(context.Background(), app, s.record())

				hasErr := err != nil
				if hasErr != s.expectError {
					t.Fatalf("Expected hasErr %v, got %v (%v)", s.expectError, hasErr, err)
				}
			})
		}
	})
}

func TestBoolFieldValidateSettings(t *testing.T) {
	t.Parallel()

	tests.DualDBTest(t, func(t *testing.T, app *tests.TestApp, dbType tests.DBType) {
		testDefaultFieldIdValidationWithApp(t, app, core.FieldTypeBool)
		testDefaultFieldNameValidationWithApp(t, app, core.FieldTypeBool)
	})
}

File: pocketless/src/core/field_bool.test.ts
/**
 * field_bool.test.ts — T137 移植 Go 版 core/field_bool_test.go
 * 对照 Go 版所有 test case
 */
import { describe, test, expect } from "bun:test";
import { BoolField } from "./field_bool";
import { RecordModel } from "./record_model";
import { CollectionModel } from "./collection_model";

function newBaseCollection(name: string): CollectionModel {
  const c = new CollectionModel();
  c.name = name;
  return c;
}

// ============================================================
// TestBoolFieldBaseMethods
// ============================================================
describe("BoolField base methods", () => {
  test("type is 'bool'", () => {
    const f = new BoolField();
    expect(f.type).toBe("bool");
  });

  test("default values", () => {
    const f = new BoolField();
    expect(f.id).toBe("");
    expect(f.name).toBe("");
    expect(f.system).toBe(false);
    expect(f.hidden).toBe(false);
    expect(f.required).toBe(false);
  });
});

// ============================================================
// TestBoolFieldColumnType
// ============================================================
describe("BoolField columnType", () => {
  test("default column type", () => {
    const f = new BoolField();
    expect(f.columnType()).toBe("BOOLEAN DEFAULT FALSE NOT NULL");
  });
});

// ============================================================
// TestBoolFieldPrepareValue — 对照 Go 版 (strconv.ParseBool)
// ============================================================
describe("BoolField prepareValue", () => {
  const f = new BoolField();

  const scenarios: { raw: unknown; expected: boolean }[] = [
    { raw: "", expected: false },
    { raw: "f", expected: false },
    { raw: "t", expected: true },
    { raw: 1, expected: true },
    { raw: 0, expected: false },
    { raw: true, expected: true },
    { raw: false, expected: false },
    { raw: "true", expected: true },
    { raw: "false", expected: false },
    { raw: "1", expected: true },
    { raw: "0", expected: false },
    { raw: "yes", expected: true },
    { raw: null, expected: false },
    { raw: undefined, expected: false },
  ];

  for (const s of scenarios) {
    test(`prepareValue(${JSON.stringify(s.raw)}) → ${s.expected}`, () => {
      const v = f.prepareValue(s.raw);
      expect(v).toBe(s.expected);
    });
  }
});

// ============================================================
// TestBoolFieldValidateValue — 对照 Go 版所有 scenario
// ============================================================
describe("BoolField validateValue", () => {
  const collection = newBaseCollection("test_collection");

  test("false field value (non-required)", () => {
    const f = new BoolField({ name: "test" });
    const record = new RecordModel(collection);
    record.set("test", false);
    const err = f.validateValue(false, record);
    expect(err).toBeNull();
  });

  test("false field value (required)", () => {
    const f = new BoolField({ name: "test", required: true });
    const record = new RecordModel(collection);
    record.set("test", false);
    const err = f.validateValue(false, record);
    expect(err).not.toBeNull();
  });

  test("true field value (required)", () => {
    const f = new BoolField({ name: "test", required: true });
    const record = new RecordModel(collection);
    record.set("test", true);
    const err = f.validateValue(true, record);
    expect(err).toBeNull();
  });
});

// ============================================================
// TestBoolFieldValidateSettings
// ============================================================
describe("BoolField validateSettings", () => {
  test("always valid", () => {
    const f = new BoolField({ id: "test", name: "test" });
    const collection = newBaseCollection("test_collection");
    expect(f.validateSettings(collection)).toBeNull();
  });
});
